From 5c1c7c52fe5d4a52679ec7eebda466f1fde7e255 Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Fri, 16 May 2025 09:32:42 +0500
Subject: [PATCH] BACKPORT: sched/fair: Iterate in ascending CPU order when
 doing idle load balance On asymmetric arm64 systems, CPUs are enumerated by
 ascending order of capacity. Since the idle load balancer doesn't iterate
 through possible CPU candidates starting from the beginning of the idle CPU
 mask, higher capacity CPUs may receive tasks before qualified lower capacity
 CPUs. This results in a small hit to energy efficiency, since higher capacity
 CPUs are typically far less efficient than lower capacity CPUs.

Iterate through the small CPUs first as possible load-balance destination
CPUs, so that tasks are more likely to be migrated to the lowest capacity
CPU that fits them, thus saving energy.

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
---
 kernel/sched/fair.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index dfc8097622d1..8b78b9713aba 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -11816,7 +11816,7 @@ static bool _nohz_idle_balance(struct rq *this_rq, unsigned int flags,
 	cpumask_andnot(&cpus, nohz.idle_cpus_mask, cpu_isolated_mask);
 
 	for_each_cpu(balance_cpu, &cpus) {
-		if (balance_cpu == this_cpu || !idle_cpu(balance_cpu))
+		if (!idle_cpu(balance_cpu))
 			continue;
 
 		/*
-- 
2.49.0

